### PR TITLE
Native support for multi-module Maven projects

### PR DESCRIPTION
**Pull request by jtnord**
This changes the way the PCT clones from per plugin to per repository.

Once the repository is cloned it then exercises each plugin in the repo in turn (serially).
plugins that are excluded (or implicitly if an include list is provided) will not be tested even if some other plugin in the same repository is tested.  It is assumed that all plugins in the same repository will use the same tag (however this is not explicitly prevented - but can be if deemed necessary)

The pct command is no longer a root command but is now a sub command `test-plugins` and any caller of the PCT will need to be adjusted to use this sub command.

There is a new command `list-plugins` that will list all plugins in the war (excluding detached) that can also filter by includes or excludes.

Tooling can call this in order to group plugin tests for the same repository.  (as above this does not group by repository and tag - it is assumed that the version of plugins from the same repository will be the same - this can be added if required, or we can group by repository and commit)

All `AbstractMultiParentHook` have been removed and similar to the `git url` workaround has been inlined into code that should become dead once all plugins have been updated to include https://github.com/jenkinsci/maven-hpi-plugin/pull/436 
The support is currently via a hook like `MetadataExtractor` - all implkementations of this are expected to be removed when plugins are migrated, simplifying the implementation.

build/testing unconditionally sets `set.changelist` so that incremental and CD versions have the correct version and can obtain parents or other reactor versions from the repository.  This is considered harmless for non incremental/CD versions.

`MavenRunner` now takes a module name which will be used as `-pl moduleName` if non null.

`--local-checkout-dir` is supported for via https://github.com/jenkinsci/maven-hpi-plugin/pull/463

This has been tested with a build against cloudbees as well as local builds with local checkouts of various plugins including `warnings-ng` and `aws-java-sdk`

`aws-java-sdk` does not have a `MetadataExtractor` and i fit is in the war is required to be version `1.12.406-373.v59d2b_d41281b_`  from https://github.com/jenkinsci/aws-java-sdk-plugin/pull/956

the `maven:hpi` for  `--local-checkout-dir` is also an incremental version which requires incremental support to be enabled when testing.

~The code may or may not support timestamped snapshots of plugins in the war. without this change they are not supported as the `tag` will be rejected due to it being `head`.  Post this change and with https://github.com/jenkinsci/maven-hpi-plugin/pull/436 timestamped snapshots in the war may work (but it has not been tested).~

The version of the core is now directly obtained from the war's Manifest, if the manifest is being discarded as part of building an uber-war then consumers will need to adapt the building of the war accordingly - an example is shown in the pom where we build our war for testing.

- [ ] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [ ] Ensure that the pull request title represents the desired changelog entry
- [ ] Please describe what you did
- [ ] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

File: src/main/java/org/jenkins/tools/test/VersionProvider.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    public String[] getVersion() throws Exception {
+        return new String[] {getPCTVersionString(), getJavaVersionString()};
+    }
+
+    @SuppressFBWarnings(
+            value = {"NP_LOAD_OF_KNOWN_NULL_VALUE", "RCN_REDUNDANT_NULLCHECK_OF_NULL_VALUE"},
+            justification =
+                    "spotbugs false poisitive due to try-with-resources / https://github.com/spotbugs/spotbugs/issues/2191")
+    private static String getPCTVersionString() {
+        try (InputStream manifestStream = VersionProvider.class.getResourceAsStream("/META-INF/MANIFEST.MF")) {
```
Comment (jtnord) [inline src/main/java/org/jenkins/tools/test/VersionProvider.java:22]: uses the metadata we now build into the jar.  unrelated to the multi-module work - but was here as it was in all the builds I had for various testing.  Should have been puled out earlier.

Comment (basil) [reply src/main/java/org/jenkins/tools/test/VersionProvider.java:22]: >  Should have been puled out earlier.

Why wasn't it?

File: src/main/java/org/jenkins/tools/test/PluginCompatTester.java [hunk 2]
```diff
@@ -93,125 +87,100 @@ public void testPlugins() throws PluginCompatibilityTesterException {
...
+                config.getExcludePlugins());
+
+        // filter any plugins that are not being tested and group by Git URL
+        // and run through the pre-checkout hooks
+        // and group by Git URL
+        Map<String, List<PluginMetadata>> pluginsByrepo;
+        try {
+            pluginsByrepo = pluginMetadataList.stream()
+                    .map(new RunAndMapBeforeCheckoutHooks(pcth, coreVersion, config))
+                    .collect(Collectors.groupingBy(PluginMetadata::getGitUrl, HashMap::new, Collectors.toList()));
```
Comment (basil) [inline src/main/java/org/jenkins/tools/test/PluginCompatTester.java:104]: Convention in this codebase is to prefer `LinkedHashMap` to `HashMap` to prefer the order specified by the user:

```suggestion
                    .collect(Collectors.groupingBy(PluginMetadata::getGitUrl, LinkedHashMap::new, Collectors.toList()));
```

File: src/main/java/org/jenkins/tools/test/PluginCompatTester.java [hunk 3]
```diff
@@ -223,141 +192,52 @@ private void testPluginAgainst(
...
         pcth.runBeforeExecution(forExecutionHooks);
 
         Map<String, String> properties = new LinkedHashMap<>(config.getMavenProperties());
         properties.put("overrideWar", config.getWar().toString());
         properties.put("jenkins.version", coreVersion);
         properties.put("useUpperBounds", "true");
+        // enable setting the incremental / CD release so if a multi-module reactor
+        // any other projects can be obtained from the repo
+        properties.put("set.changelist", "true");
+        // as hooks may be adjusting the POMs tell the git-changelist extension to ignore dirty
```
Comment (basil) [inline src/main/java/org/jenkins/tools/test/PluginCompatTester.java:238]: :eyes:

File: src/main/java/org/jenkins/tools/test/PluginCompatTester.java [hunk 4]
```diff
@@ -661,130 +530,107 @@ private static void cloneImpl(String gitUrl, String scmTag, File checkoutDirecto
...
+     * @param <T>
+     * @param current the PluginCompatibilityTesterException if any
+     * @param caught the newly caught exception
+     * @param throwException {@code true} if we should immediatly rethrow {@code caught}, {@code
+     *     false} indicating we should return {@caught}.
+     * @return {@code caught}
+     * @throws PluginCompatibilityTesterException if {@code ethrowException == true} then {@caught}
+     *     is thrown.
+     */
+    private static <T extends PluginCompatibilityTesterException> T throwOrAddSupressed(
```
Comment (basil) [inline src/main/java/org/jenkins/tools/test/PluginCompatTester.java]: This is a nice refactoring but is unnecessarily cluttering up this PR. Can we file a PR against the main branch to implement this cleanup there so that this diff shrinks and is easier to understand for readers?

File: src/main/java/org/jenkins/tools/test/exception/WrappedPluginCompatibilityException.java [hunk 5]
```diff
@@ -0,0 +1,17 @@
+package org.jenkins.tools.test.exception;
+
+/** A {@code RuntimeException} that wraps a PluginCompatibilityTesterException. */
+public class WrappedPluginCompatibilityException extends RuntimeException {
```
Comment (basil) [inline src/main/java/org/jenkins/tools/test/exception/WrappedPluginCompatibilityException.java:4]: [nit] This is like the JDK's `UncheckedIOException`, so I think it would be nice to use a similar name here e.g. `UncheckedPluginCompatibilityTesterException`.

File: src/main/java/org/jenkins/tools/test/model/hook/BeforeCompilationContext.java [hunk 6]
```diff
@@ -3,47 +3,25 @@
...
-        return parentFolder;
-    }
-
-    public boolean ranCompile() {
-        return ranCompile;
-    }
-
-    public void setRanCompile(boolean ranCompile) {
-        this.ranCompile = ranCompile;
+    public File getCloneDir() {
```
Comment (basil) [inline src/main/java/org/jenkins/tools/test/model/hook/BeforeCompilationContext.java:24]: Naming is inconsistent with the field name `cloneDirectory`.

File: src/main/java/org/jenkins/tools/test/model/hook/HookOrderComparator.java [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+import java.io.Serializable;
+import java.util.Comparator;
+
+/**
+ * Comparator that will order first based on the {@code @HookOrder}. Objects whose classes are
+ * missing the annotation order will be treated as if they have the default order. Where two objects
+ * have the same order sorting will occur based on the classname.
+ */
+public class HookOrderComparator implements Comparator<Object>, Serializable {
```
Comment (basil) [inline src/main/java/org/jenkins/tools/test/model/hook/HookOrderComparator.java:11]: This seems useful but is cluttering up the diff with an unrelated change. Can we file a separate PR for the hook order feature, with some unit tests, in order to keep this diff smaller and more manageable for the reader?

File: src/main/java/org/jenkins/tools/test/model/plugin_metadata/PluginMetadataHooks.java [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+        List<URL> urls = new ArrayList<>();
+        for (File jar : externalJars) {
+            try {
+                urls.add(jar.toURI().toURL());
+            } catch (MalformedURLException e) {
+                throw new UncheckedIOException(e);
+            }
+        }
+        return new URLClassLoader(urls.toArray(new URL[0]), base);
+    }
```
Comment (basil) [inline src/main/java/org/jenkins/tools/test/model/plugin_metadata/PluginMetadataHooks.java:56]: This entails a bit of duplication with `PluginCompatTesterHooks`; can it be de-duped?

File: src/main/java/org/jenkins/tools/test/util/ModelReader.java [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+        String pom = getPomFromHpi(groupId, artifactId, jarInputStream);
+        Model model;
+
+        try (Reader r = new StringReader(pom)) {
+            MavenXpp3Reader mavenXpp3Reader = new MavenXpp3Reader();
+            model = mavenXpp3Reader.read(r);
+        } catch (XmlPullParserException e) {
+            throw new PluginSourcesUnavailableException("Failed to parse pom.xml", e);
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
```
Comment (basil) [inline src/main/java/org/jenkins/tools/test/util/ModelReader.java:41]: If this method is throwing `IOException` then you don't need this catch block.

File: src/main/java/org/jenkins/tools/test/PluginListerCli.java [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+        return Integer.valueOf(0);
+    }
+
+    private static String formatEntry(Entry<String, List<PluginMetadata>> entry) {
+        StringBuilder sb = new StringBuilder(entry.getKey());
+        for (PluginMetadata pm : entry.getValue()) {
+            sb.append("\t").append(pm.getPluginId());
+        }
+        return sb.toString();
+    }
```
Comment (basil) [inline src/main/java/org/jenkins/tools/test/PluginListerCli.java:116]: Interactive output of `list-plugins` is unreadable:

```
https://github.com/jenkinsci/workflow-multibranch-plugin.git    workflow-multibranch
https://github.com/jenkinsci/jakarta-activation-api-plugin.git  jakarta-activation-api
https://github.com/jenkinsci/text-finder-plugin.git     text-finder
https://github.com/jenkinsci/file-parameters-plugin.git file-parameters
https://github.com/jenkinsci/cloudbees-folder-plugin.git        cloudbees-folder
https://github.com/jenkinsci/github-branch-source-plugin.git    github-branch-source
https://github.com/jenkinsci/workflow-basic-steps-plugin.git    workflow-basic-steps
https://github.com/jenkinsci/ws-cleanup-plugin.git      ws-cleanup
https://github.com/jenkinsci/jersey2-api-plugin.git     jersey2-api
https://github.com/jenkinsci/instance-identity-plugin.git       instance-identity
https://github.com/jenkinsci/mina-sshd-api-plugin.git   mina-sshd-api-sftp      mina-sshd-api-scp       mina-sshd-api-common    mina-sshd-api-core
https://github.com/jenkinsci/javax-mail-api-plugin.git  javax-mail-api
https://github.com/jenkinsci/ionicons-api-plugin        ionicons-api
https://github.com/jenkinsci/script-security-plugin.git script-security
https://github.com/jenkinsci/copyartifact-plugin.git    copyartifact
https://github.com/jenkinsci/slack-plugin       slack
https://github.com/jenkinsci/ant-plugin.git     ant
https://github.com/jenkinsci/jackson2-api-plugin.git    jackson2-api
https://github.com/jenkinsci/aws-java-sdk-plugin.git    aws-java-sdk-elasticbeanstalk   aws-java-sdk-cloudformation     aws-java-sdk-sqs        aws-java-sdk-ssm        aws-java-sdk-codebuild  aws-java-sdk-logs       aws-java-sdk-efs      aws-java-sdk-iam        aws-java-sdk-minimal    aws-java-sdk-ecr        aws-java-sdk-ec2        aws-java-sdk-ecs        aws-java-sdk-sns
```

For it to be readable, it should have a heading and clearly defined column names. For example:

```
PLUGIN                    URL
jakarta-activation-api    https://github.com/jenkinsci/jakarta-activation-api-plugin.git
mina-sshd-api-common      https://github.com/jenkinsci/mina-sshd-api-plugin.git
mina-sshd-api-core        https://github.com/jenkinsci/mina-sshd-api-plugin.git
mina-sshd-api-scp         https://github.com/jenkinsci/mina-sshd-api-plugin.git
mina-sshd-api-sftp        https://github.com/jenkinsci/mina-sshd-api-plugin.git
workflow-multibranch      https://github.com/jenkinsci/workflow-multibranch-plugin.git
```

I do not care how the machine-parsable version looks, as long as it is clearly documented and consumers can rely on the machine-parsable format being stable.

File: src/main/java/org/jenkins/tools/test/model/plugin_metadata/LocalCheckoutMetadataExtractor.java [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+        try {
+            runner.run(
+                    Map.of("output", log.getAbsolutePath()),
+                    localCheckoutDir,
+                    null,
+                    null,
+                    "-q",
+                    // TODO to switch to release version
+                    // https://github.com/jenkinsci/maven-hpi-plugin/pull/463
+                    "org.jenkins-ci.tools:maven-hpi-plugin:3.42-rc1408.71cefb_fc63b_d:list-plugins",
```
Comment (basil) [inline src/main/java/org/jenkins/tools/test/model/plugin_metadata/LocalCheckoutMetadataExtractor.java:48]: I think we should override this version only if necessary (i.e., only if the plugin is using a version older than the minimum required version). There is code elsewhere in this repository that implements this already.

File: pom.xml [hunk 12]
```diff
@@ -256,14 +256,47 @@
                 <delete dir="${project.build.directory}/megawar/WEB-INF/detached-plugins" />
                 <delete file="${project.build.directory}/megawar/META-INF/JENKINS.RSA" />
                 <delete file="${project.build.directory}/megawar/META-INF/JENKINS.SF" />
-                <war destfile="${project.build.directory}/megawar.war">
+                <war destfile="${project.build.directory}/megawar.war" manifest="${project.build.directory}/megawar/META-INF/MANIFEST.MF">
```
[the remark this answers is not in the document]
Comment (basil) [reply pom.xml]: Pulled out into #515.

File: pom.xml [hunk 13]
```diff
@@ -256,14 +256,47 @@
...
                   <fileset dir="${project.build.directory}/megawar" />
                 </war>
               </target>
             </configuration>
           </execution>
         </executions>
       </plugin>
+      <plugin>
```
[the remark this answers is not in the document]
Comment (basil) [reply pom.xml]: Pulled out into #517.

File: src/main/java/org/jenkins/tools/test/PluginCompatTester.java (hunk 3 above)
Comment (jtnord) [reply src/main/java/org/jenkins/tools/test/PluginCompatTester.java:238]: it is legal (there are no requirements to say otherwise) and whilst the OSS ones have been re-written to use properties - some proprietary ones exist that manipulate the pom (e.g. CloudBees' wars have an extra jar in `WEB-INF/lib` of the war that we need to update for plugins that depend on it and as it is not in OSS Jenkins nor is a plugin the `maven-hpi-plugin` is unaware of it.

File: src/main/java/org/jenkins/tools/test/PluginCompatTester.java (hunk 4 above)
Comment (jtnord) [reply src/main/java/org/jenkins/tools/test/PluginCompatTester.java]: sure.

File: src/main/java/org/jenkins/tools/test/model/hook/HookOrderComparator.java (hunk 7 above)
Comment (jtnord) [reply src/main/java/org/jenkins/tools/test/model/hook/HookOrderComparator.java:11]: sure

File: src/main/java/org/jenkins/tools/test/PluginCompatTester.java [hunk 14]
```diff
@@ -94,281 +87,159 @@ public void testPlugins() throws PluginCompatibilityTesterException {
                 new PluginCompatTesterHooks(config.getExternalHooksJars(), config.getExcludeHooks());
         // Determine the plugin data
 
-        // Scan bundled plugins. If there is any bundled plugin, only these plugins will be taken
-        // under the consideration for the PCT run.
-        UpdateSite.Data data = scanWAR(config.getWar(), "WEB-INF/(?:optional-)?plugins/([^/.]+)[.][hj]pi");
```
[the remark this answers is not in the document]
Comment (basil) [reply src/main/java/org/jenkins/tools/test/PluginCompatTester.java:98]: Nice!

File: src/main/java/org/jenkins/tools/test/util/ModelReader.java (hunk 9 above)
Comment (basil) [reply src/main/java/org/jenkins/tools/test/util/ModelReader.java]: Fixed in your branch

File: src/main/java/org/jenkins/tools/test/hook/TagValidationHook.java [hunk 15]
```diff
@@ -17,9 +16,10 @@ public boolean check(@NonNull BeforeCheckoutContext context) {
 
     @Override
     public void action(@NonNull BeforeCheckoutContext context) throws PluginSourcesUnavailableException {
-        Model model = context.getModel();
-        if (model.getScm().getTag() == null || model.getScm().getTag().equals("HEAD")) {
-            throw new PluginSourcesUnavailableException("Failed to check out plugin sources for " + model.getVersion());
+        String tag = context.getPluginMetadata().getTag();
+        if (tag == null || tag.equals("HEAD")) {
```
Comment (jtnord) [inline src/main/java/org/jenkins/tools/test/hook/TagValidationHook.java:20]: this is incorrect.
we should only abort if we do not know a version in order to update all plugins and checkout out.

the whole point of using the commit hash and not the tag is that so this can work with timestamped snapshots.

As far as the checkout is concerned - using the hash is correct.
if the `hpi-plugin` can not cope with timestamped snapshots when updating dependencies then that is a bug I will need to fix also, however as it stands this regresses support for timestamped snapshots.

File: src/main/java/org/jenkins/tools/test/PluginListerCli.java (hunk 10 above)
Comment (jtnord) [reply src/main/java/org/jenkins/tools/test/PluginListerCli.java]: addressed in 8fad52370bf793321db77fe286792695efc960ea

output now uses spaces for consistent padding and follows the following example.

```
Apr 12, 2023 8:06:02 PM org.jenkins.tools.test.model.plugin_metadata.PluginMetadataHooks getPluginDetails
INFO: Extracting metadata about workflow-step-api
Apr 12, 2023 8:06:02 PM org.jenkins.tools.test.model.plugin_metadata.PluginMetadataHooks getPluginDetails
INFO: Extracting metadata about workflow-support
Plugin ID                                     Git Repository URL
ace-editor                                    https://github.com/jenkinsci/js-libs.git
active-directory                              https://github.com/jenkinsci/active-directory-plugin.git
analysis-model-api                            https://github.com/jenkinsci/analysis-model-api-plugin.git
ant                                           https://github.com/jenkinsci/ant-plugin.git
antisamy-markup-formatter                     https://github.com/jenkinsci/antisamy-markup-formatter-plugin.git
apache-httpcomponents-client-4-api            https://github.com/jenkinsci/apache-httpcomponents-client-4-api-plugin.git
artifact-manager-s3                           https://github.com/jenkinsci/artifact-manager-s3-plugin.git
authentication-tokens                         git://github.com/jenkinsci/authentication-tokens-plugin.git
aws-credentials                               git://github.com/jenkinsci/aws-credentials-plugin.git
aws-global-configuration                      git://github.com/jenkinsci/aws-global-configuration-plugin.git
aws-java-sdk-ec2                              https://github.com/jenkinsci/aws-java-sdk-plugin.git
aws-java-sdk-elasticbeanstalk                 https://github.com/jenkinsci/aws-java-sdk-plugin.git
aws-java-sdk-minimal                          https://github.com/jenkinsci/aws-java-sdk-plugin.git
blueocean-bitbucket-pipeline                  https://github.com/jenkinsci/blueocean-plugin.git
blueocean-commons                             https://github.com/jenkinsci/blueocean-plugin.git
blueocean-config                              https://github.com/jenkinsci/blueocean-plugin.git
blueocean-core-js                             https://github.com/jenkinsci/blueocean-plugin.git
blueocean-dashboard                           https://github.com/jenkinsci/blueocean-plugin.git
blueocean-display-url                         git://github.com/jenkinsci/blueocean-display-url-plugin.git
blueocean-events                              https://github.com/jenkinsci/blueocean-plugin.git
blueocean-git-pipeline                        https://github.com/jenkinsci/blueocean-plugin.git
```

File: src/main/java/org/jenkins/tools/test/model/plugin_metadata/PluginMetadataExtractor.java [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+
+    /**
+     * Obtain the metadata for a give plugin from a jar file.
+     *
+     * @param manifest the plugins' manifest.
+     * @param model the plugins' model (from the HPI).
+     * @return a fully populated PluginMetadata (or empty) for the given plugin.
+     */
+    public abstract Optional<PluginMetadata> extractMetadata(String pluginId, Manifest manifest, Model model)
+            throws MetadataExtractionException;
```
[the remark this answers is not in the document]
Comment (jtnord) [reply src/main/java/org/jenkins/tools/test/model/plugin_metadata/PluginMetadataExtractor.java:21]: I could do this - at the same time we should not be introducing any other extractors other than what is here and in proprietary (CB MultiModules) and these should themselves be removed once we have everything using the new hpi-plugin.

Additionally what would `isApplicable` return for the `LegacyPluginMetadateExtractor` `true` unless the plugin is missing the SCM - but order is important here - it is not applicable to any multi module plugin so having the implantation return `true` would be seemingly incorrect and confusing also - even if it was never called.

The exception case is no longer used - and so would removing the throws declaration make things clearer here?

le me know your thoughts and I will adjust as suitable.

File: src/main/java/org/jenkins/tools/test/model/plugin_metadata/PluginMetadataHooks.java (hunk 8 above)
Comment (jtnord) [reply src/main/java/org/jenkins/tools/test/model/plugin_metadata/PluginMetadataHooks.java:56]: it can.  I saw this whole class as temporary to be removed once everything had adjusted to the new parent pom, but can extract the commonality for sure.

File: src/main/java/org/jenkins/tools/test/model/hook/HookOrderComparator.java (hunk 7 above)
Comment (jtnord) [reply src/main/java/org/jenkins/tools/test/model/hook/HookOrderComparator.java]: #521

File: src/main/java/org/jenkins/tools/test/PluginListerCLI.java [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+
+        if (output != null) {
+            try (BufferedWriter writer = Files.newBufferedWriter(
+                    output.toPath(),
+                    StandardOpenOption.CREATE,
+                    StandardOpenOption.TRUNCATE_EXISTING,
+                    StandardOpenOption.WRITE)) {
+                for (Map.Entry<String, List<PluginMetadata>> entry : metaDataByRepoMap.entrySet()) {
+                    writer.write(formatEntry(entry));
+                    writer.write("\\n");
```
[the remark this answers is not in the document]
Comment (basil) [reply src/main/java/org/jenkins/tools/test/PluginListerCLI.java]: I fixed this.

File: src/main/java/org/jenkins/tools/test/exception/WrappedPluginCompatibilityException.java (hunk 5 above)
Comment (basil) [reply src/main/java/org/jenkins/tools/test/exception/WrappedPluginCompatibilityException.java]: I got rid of this class entirely in one of my commits.

File: src/main/java/org/jenkins/tools/test/model/hook/BeforeCompilationContext.java (hunk 6 above)
Comment (basil) [reply src/main/java/org/jenkins/tools/test/model/hook/BeforeCompilationContext.java]: I fixed this.

File: src/main/java/org/jenkins/tools/test/PluginListerCli.java (hunk 10 above)
Comment (basil) [reply src/main/java/org/jenkins/tools/test/PluginListerCli.java]: I cleaned this up even further.

**Comment by basil** [comment]
Latest revision passing `jenkinsci/bom` at https://github.com/jenkinsci/bom/pull/1961 :+1:

**Review by basil — APPROVED** [review]
Approval conditional on:

- `loadExtractors` and `setupExternalClassLoaders` are deduplicated with the versions in `PluginCompatTesterHooks`
- The `throwOrAddSuppressed` is checked in to the main branch and merged back in to shrink the diff
- The `getFallbackGitUrl` cleanup is checked in to the main branch and merged back in to shrink the diff

**Comment by basil** [comment]
Please do not retain any of my sleep-deprived late-night commit messages (notes to self, not intended for the historical record) by squash merging this, retaining only the title and Co-authored-by.

**Comment by jtnord** [comment]
I think I have regressed something in the recent changes, possibly service loader.  looking at it now.

**Comment by jtnord** [comment]
> There should not be any external extractors, and the presence of an external extractor would be technical debt that would prevent us from simplifying this logic in the future.

perhaps you are forgetting about the multi-module proprietary plugins?

e.g. we have an extractor similar in nature to [`LegacyMultiModulePluginMetadataExtractor`](https://github.com/jenkinsci/plugin-compat-tester/blob/505510ca47d29772444215d75585e41dfc8077f4/src/main/java/org/jenkins/tools/test/model/plugin_metadata/LegacyMultiModulePluginMetadataExtractor.java)

Or are you suggesting we sit on this code until all required plugins have been updated?

**Comment by basil** [comment]
That could be OK. Can you share a link to what you have in mind?

File: src/main/java/org/jenkins/tools/test/util/WarExtractor.java [hunk 18]
```diff
@@ -0,0 +1,160 @@
...
+                if (isInteresting(entry)) {
+                    plugins.add(getPlugin(jf, entry));
+                }
+            }
+        } catch (IOException e) {
+            throw new UncheckedIOException("I/O error occurred whilst extracting plugin metadata from WAR", e);
+        }
+        if (plugins.isEmpty()) {
+            throw new MetadataExtractionException("Found no plugins in " + warFile);
+        }
```
Comment (jtnord) [inline src/main/java/org/jenkins/tools/test/util/WarExtractor.java:93]: this breaks the ability to run a PCT against a plugin that is in a local clone only.  this was previously working,
